lib/phoenix_component.ex

```diff
@@ -2014,6 +2014,128 @@ defmodule Phoenix.Component do
+  )
+
+  attr.(:id, :string,
+    doc: """
+    The id to be used in the form, defaults to the concatenation of the given
+    field to the parent form id.
+    """
+  )
+
+  attr.(:as, :atom,
+    doc: """
+    The name to be used in the form, defaults to the concatenation of the given
+    field to the parent form name.
+    """
+  )
+
+  attr.(:default, :any, doc: "The value to use if none is available.")
+
+  attr.(:prepend, :list,
+    doc: """
+    The values to prepend when rendering. This only applies if the field value
+    is a list and no parameters were sent through the form.
+    """
+  )
+
+  attr.(:append, :list,
+    doc: """
+    The values to append when rendering. This only applies if the field value
+    is a list and no parameters were sent through the form.
+    """
+  )
+
+  attr.(:skip_hidden, :boolean,
+    default: false,
+    doc: """
+    Skip the automatic rendering of hidden fields to allow for more tight control
+    over the generated markup. You can access `form.hidden` or use `.hidden_inputs/1`
+    to generate them manually.
+    """
+  )
+
+  slot.(:inner_block, required: true, doc: "The content rendered for each nested form.")
+
+  def inputs_for(assigns) do
+    {form, field} = assigns[:field] || raise ArgumentError, "missing :field assign to inputs_for"
+
+    options =
+      assigns
+      |> Map.take([:id, :as, :default, :append, :prepend])
+      |> Enum.reject(fn {_, v} -> v == nil end)
+
+    options =
+      form.options
+      |> Keyword.take([:multipart])
+      |> Keyword.merge(options)
+
+    assigns =
+      assigns
+      |> assign(:field, nil)
+      |> assign(:forms, form.impl.to_form(form.source, form, field, options))
+
+    ~H"""
+    <%= for finner <- @forms do %>
+      <%= unless @skip_hidden do %>
+        <%= for {name, value_or_values} <- finner.hidden,
+                name = name_for_value_or_values(finner, name, value_or_values),
+                value <- List.wrap(value_or_values) do %>
+          <input type="hidden" name={name} value={value} />
+        <% end %>
+      <% end %>
+      <%= render_slot(@inner_block, finner) %>
+    <% end %>
+    """
+  end
+
+  @spec name_for_value_or_values(Phoenix.HTML.Form.t() | atom, atom | String.t(), term) ::
+          String.t()
+  defp name_for_value_or_values(form, field, values) when is_list(values) do
+    input_name(form, field) <> "[]"
+  end
+
+  defp name_for_value_or_values(form, field, _value) do
+    input_name(form, field)
+  end
+
+  @spec input_name(Phoenix.HTML.Form.t() | atom, atom | String.t()) :: String.t()
+  defp input_name(form_or_name, field)
+
+  defp input_name(%{name: nil}, field), do: to_string(field)
+
+  defp input_name(%{name: name}, field) when is_atom(field) or is_binary(field),
+    do: "#{name}[#{field}]"
+
+  defp input_name(name, field) when (is_atom(name) and is_atom(field)) or is_binary(field),
+    do: "#{name}[#{field}]"
```

Comment:
```suggestion
    Phoenix.HTML.Form.input_name(form, field) <> "[]"
  end

  defp name_for_value_or_values(form, field, _value) do
    Phoenix.HTML.Form.input_name(form, field)
  end
```

It is fine to use `input_name`.

Long story short: The goal is to release Phoenix.HTML v4.0 eventually in the future with only the building blocks and protocols. The current helpers will be moved to `PhoenixHTMLHelpers`. We don't know exactly how it will look like but I can say with 99% confidence that `input_name` will be part of the building blocks.